internal/provider/resource_gitlab_runner_test.go

```diff
@@ -0,0 +1,121 @@
+package provider
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/xanzy/go-gitlab"
+)
+
+func TestAccGitlabRunner_basic(t *testing.T) {
+	testAccCheck(t)
+
+	group := testAccCreateGroups(t, 1)[0]
+	//The runner token is not populated on the return from the group create, so re-retrieve it to get the token.
+	group, _, err := testGitlabClient.Groups.GetGroup(group.ID, &gitlab.GetGroupOptions{})
+	if err != nil {
+		t.Fail()
+	}
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() {},
+		ProviderFactories: providerFactories,
+		CheckDestroy:      testAccCheckManagedLicenseDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: fmt.Sprintf(`
+				resource "gitlab_runner" "this" {
+					token = "%s"
+					description = "Lorem Ipsum"
+				}
+				`, group.RunnersToken),
+			},
+			{
+				ResourceName:      "gitlab_runner.this",
+				ImportState:       true,
+				ImportStateVerify: true,
+				//These need to be ignored since they don't come back in the "get" command
+				ImportStateVerifyIgnore: []string{"authentication_token", "token"},
+			},
+			{
+				Config: fmt.Sprintf(`
+				resource "gitlab_runner" "this" {
+					token = "%s"
+					description = "Lorem Ipsum Dolor Sit Amet"
+				}
+				`, group.RunnersToken),
+			},
+			{
+				ResourceName:      "gitlab_runner.this",
+				ImportState:       true,
+				ImportStateVerify: true,
+				//These need to be ignored since they don't come back in the "get" command
+				ImportStateVerifyIgnore: []string{"authentication_token", "token"},
+			},
```

Comment:
Should we also test instance level runners?

Reply:
@timofurrer  - Good call! Added a test 😄

Reply:
So this test is... pretty weird. I added a pretty lengthy comment at the top of it, but there seems to be a built-in delay in how often we can rotate the instance-level token. Since the CI should run against a fresh version of gitlab, it _shouldn't_ be a problem in the CI, but we'll see...

Reply:
OK, the test is failing in the CI/CD. I'm kinda feeling like it's flakey enough that we should remove the test until we have a better way of setting the instance runner token. Thoughts?

Reply:
hmmm, so what's the problem here in the CI? Do you have the test somewhere?

Reply:
@timofurrer  - You can see the test here: https://github.com/gitlabhq/terraform-provider-gitlab/pull/1049/files#diff-3bac8c8a2d8b8ebe1b74562e5f710b5fbc9d861fbfe8ac15d00d16721481d383R67

What's failing (assuming it's the same 403 I got when I was running it locally) requires a bit of explanation: There isn't a way to retrieve the instance-level runner token via the API that I can find, which means that instead of retrieving what's there, I call a _reset_ via the API. The reset API returns the _new_ registration token, however it's not immediately available for use. If you turn around and call the API using the new registration token immediately, you receive a 403 (which is what the CI/CD is receiving). However, what makes it janky is that the amount of time necessary to wait before the token can be used seems to be variable - I've seen everything from _instantly_ to _almost a minute_ in local testing. The time it takes for it to be usable seems to depend on how often you're calling the reset API, so I was hoping that since we only called it once during acceptance testing that it would work properly. That doesn't seem to be the case :-/

I don't really want to add a really long wait to the CI/CD to account for the case... I could add a new client-based call at the top of the test that calls "resource.Retry" until we get a non-403, but I feel like we still have a pseudo-race condition here.

Reply:
Okay, yeah - I don't want to have such a delay in there either ...

Let's leave it out for now then. Do you mind adding this additional case to: https://gitlab.com/gitlab-org/gitlab/-/issues/352299

Reply:
hmmm, is there a way to inject an instance token manually like we do with the PAT we use to authenticate to the instance?

Reply:
That's an interesting question - let me do some research. Since I use SaaS, I don't have a ton of experience with the ruby console commands, but I'll dig and see. What I'll do is remove the test for now, and we can re-add it with a PR once we can figure that out.

As to the gitlab issue, that seems to be related specifically to groups/subgroups (and is assigned as such) whereas this is related to application settings proper: @nagyv  - Do you think this should go in there, or should I create a new issue and let it get applied to a different gitlab team?

Reply:
So I _may_ have found a way to make this work. Digging into the code, it looks like this is likely a caching problem (which makes sense with the way the failures exhibit): https://gitlab.com/gitlab-org/gitlab/-/blob/master/ee/app/services/ee/ci/runners/reset_registration_token_service.rb#L22

However, digging into the specs that GitLab itself uses to test the API, there is a hint there. GitLab's tests stub in an environment variable that, based on its naming, would disable that in-memory caching: https://gitlab.com/gitlab-org/gitlab/-/blob/master/spec/requests/api/ci/runners_reset_registration_token_spec.rb#L77

I'm attempting to add that to the docker-compose and will see if that resolves the issue with repeated running of the spec. I did find out how to _read_ the value from the rails console, but not how to _set_ the value from the rails console. Some of that is likely my inexperience with ruby as a language. We _could_ read the value into a file, then parse the file from the test, but that seems like a really janky workaround. I hope the above works.

Reply:
Even better - we can set it in the gitlab.rb: https://gitlab.com/gitlab-org/omnibus-gitlab/blob/master/files/gitlab-config-template/gitlab.rb.template#L687

I'm going to open a new PR that adds that so we can merge that and then rebase for this PR.

Reply:
Yes - that's exactly what I was hoping will work 🎉

Reply:
@timofurrer  - I've rebased to master (and squashed for a clean history) and re-added the test. I ran it in gitpod several times in a row with success. Looks like we're clean here!

Reply:
Great work @PatrickRice-KSC 🎉